### DISCUSSION TITLE
Autoscaling KafkaConnectors

### DISCUSSION THREAD

**Discussion by opsocomusr** [strimzi/strimzi-kafka-operator#6871] [Q&A] [answered]
We are trying to setup autoscaling for KafkaConnector resources using KEDA. However when we deploy the scaled object, we are getting the following error in the HPA.

```
KafkaConnector pods:  5 current / 0 desired
Conditions:
  Type           Status  Reason             Message
  ----           ------  ------             -------
  AbleToScale    True    SucceededGetScale  the HPA controller was able to get the target's current scale
  ScalingActive  False   InvalidSelector    the HPA target's scale is missing a selector
Events:
  Type     Reason                        Age                   From                       Message
  ----     ------                        ----                  ----                       -------
  Warning  FailedComputeMetricsReplicas  19m (x12 over 22m)    horizontal-pod-autoscaler  selector is required
  Warning  SelectorRequired              2m39s (x81 over 22m)  horizontal-pod-autoscaler  selector is required
```

In the [KafkaConnector CRD](https://github.com/strimzi/strimzi-kafka-operator/blob/main/helm-charts/helm3/strimzi-kafka-operator/crds/047-Crd-kafkaconnector.yaml#L29-L31), there is no labelSelectorPath in the scale object.

Is autoscaling not supported yet for KafkaConnector resources? From the other [discussion](https://github.com/strimzi/strimzi-kafka-operator/discussions/6635#discussioncomment-2514171) comment, we were under assumption that this is supported for KafkaConnectors.

**Comment by scholzj** [comment, marked as answer]
The `scale` subresource is supported by the KafkaConnector resource. But the KafkaConnector is a virtual object which does not represent any pods. So there is nothing the label selector can select.

**Reply by opsocomusr** [reply]
So does that mean we cannot use HPA to autoscale KafkaConnector?

`kubectl scale kctr <KAFKA-CONNECTOR-NAME> --replicas=5` works as expected.

**Reply by scholzj** [reply]
I think that means you cannot use HPA since HPA is IIRC designed around pods.

**Reply by arthurh4** [reply]
Hello @scholzj,

Would it be do-able to add the selector field to status.selector in scale call ? like following https://medium.com/@thescott111/autoscaling-kubernetes-custom-resource-using-the-hpa-957d00bb7993

I verify with strimzi:
`kubectl get --raw /apis/kafka.strimzi.io/v1beta2/namespaces/MY_NAMESPACE/kafkaconnects/MY_CONNECT/scale`

Output:
  "spec": {
    "replicas": 4
  },
  "status": {
    "replicas": 4
  }
If i understand we need to add the pod selector on the status field to make it work.

What's confusing me if that the code should return that value no ? https://github.com/strimzi/strimzi-kafka-operator/blob/main/helm-charts/helm3/strimzi-kafka-operator/crds/041-Crd-kafkaconnect.yaml#L32 and https://github.com/strimzi/strimzi-kafka-operator/blob/5fb901386818a4eea25cc0c5618cb45204f84363/api/src/main/java/io/strimzi/api/kafka/model/status/KafkaConnectStatus.java#L35

i'm trying to look in the code how this api call is handled

**Reply by scholzj** [reply]
The problem is that there is no selector for the Connector tasks. They are not Kubernetes objects. And we cannot point it to the Connect pods because the number will not match. The number of tasks the connector has has no correlation to the number of Connect pods in Kubernetes. You can have 20 tasks on 4 pods. Or you can have 2 tasks on top of Kafka Connect with 5 pods (because it runs also other connectors and tasks).

**Reply by arthurh4** [reply]
my bad, i did not see this was about kafkaconnectors, I'm actualy trying to scale other objects (kafka,kafkaconnect,mirrormakers) and seem to encounter the same issue.
Are other types supposed to me auto scalable ?

**Reply by scholzj** [reply]
It should be possible to use HPA for Connect and Mirror Makers. They have the scale subresources as well as the selector. I for sure used it with Connect. For Kafka, we currently do not support the scale subresource ... scaling Kafka brokers is kind of complicated on its own even without auto-scaling.

**Reply by hafizmujadidKhalid** [reply]
@scholzj!
Can you please point me to some example of how did you manage autoscaling for connect? thanks

**Reply by scholzj** [reply]
I do not have any examples to share. But IIRC, you simply just set-up the `HorizontalPodAutoscaler` resource and point it to the `KafkaConnect` resource.

**Reply by hafizmujadidKhalid** [reply]
@scholzj! thanks for the hint. So, it will scale connect workers, can we do something about connector tasks as well or is it something not possible? I am not an expert in K8s, excuse me for my ignorance :)

**Reply by scholzj** [reply]
As discussed earlier here -> the Connector tasks have the scale endpoint, bu they do not have any Pod representation. So I'm not sure if regular HPA resource would work on it. 

Also, scaling the connectors is a it more complicated => the `tasks.max` option does not mean how many tasks will there be. Its just the maximum. Some connector will always use only one task regardless the configuration. Scalability of sink connectors will also depend on on other factors such as how many partitions does the Kafka topic has. Source connectors have often similar limitations imposed from the source system. So it is fairly tricky to have some generic autoscaling for this. You might be able to get something working for some particular connector if you understand how it works.

**Reply by hafizmujadidKhalid** [reply]
makes sense, Thanks a lot
